src/libstore/unix/builtins/unpack-channel.cc

```diff
@@ -21,10 +21,12 @@ void builtinUnpackChannel(
 
     unpackTarfile(src, out);
 
-    auto entries = readDirectory(out);
-    if (entries.size() != 1)
+    auto entries = std::filesystem::directory_iterator{out};
+    auto file_count = std::distance(entries, std::filesystem::directory_iterator{});
```

Comment:
I'm not sure if this is well-defined behaviour. `std::distance` will advance a copy of the `entries` iterator, reading the directory and (presumably) discarding the entries read. So I'm not sure if you can still read from `entries` after that.

There is a failure in `tests/functional/nix-channel.sh` that might be related to this.

Reply:
Thanks for the help. This is fixed in https://github.com/NixOS/nix/pull/10684/commits/4d0777ca69463cbaba3501c10d16012101bb0933 and tests are now pasing!